lib/harmony/harmonytask/harmonytask.go

```diff
@@ -0,0 +1,462 @@
+	e := &TaskEngine{
+		ctx:            ctx,
+		grace:          grace,
+		db:             db,
+		resources:      reg.Resources,
+		ownerID:        reg.Resources.MachineID,
+		hostAndPort:    hostnameAndPort,
+		workAdderMutex: &notifyingMx{},
+		taskMap:        make(map[string]*taskTypeHandler, len(impls)),
+		tryAllWork:     make(chan bool),
+		follows:        make(map[string][]followStruct),
+	}
+	for _, c := range impls {
+		h := taskTypeHandler{
+			TaskInterface:   c,
+			TaskTypeDetails: c.TypeDetails(),
+			TaskEngine:      e,
+		}
+		h.LastCleanup.Store(time.Now())
+		e.handlers = append(e.handlers, &h)
+		e.myTasks = append(e.myTasks, h.TaskTypeDetails.Name)
+		e.taskMap[h.TaskTypeDetails.Name] = &h
+
+		_, err := db.Exec(e.ctx, `INSERT INTO harmony_task_impl (owner_id, name) 
+			VALUES ($1,$2)`, e.ownerID, h.Name)
+		if err != nil {
+			return nil, fmt.Errorf("can't update impl: %w", err)
+		}
+
+		for name, fn := range c.TypeDetails().Follows {
+			e.follows[name] = append(e.follows[name], followStruct{fn, &h})
+
+			// populate harmony_task_follows
+			_, err := db.Exec(e.ctx, `INSERT INTO harmony_task_follows (owner_id, from_task, to_task)
+				VALUES ($1,$2,$3)`, e.ownerID, name, h.Name)
+			if err != nil {
+				return nil, fmt.Errorf("can't update harmony_task_follows: %w", err)
+			}
+		}
+	}
+
+	// resurrect old work
+	{
+		var taskRet []struct {
+			ID   int
+			Name string
+		}
+
+		err := db.Select(e.ctx, &taskRet, `SELECT id, name from harmony_task WHERE owner_id=$1`, e.ownerID)
+		if err != nil {
+			return nil, err
+		}
+		for _, w := range taskRet {
+			// edge-case: if old assignments are not available tasks, unlock them.
+			h := e.taskMap[w.Name]
+			if h == nil {
+				_, err := db.Exec(e.ctx, `UPDATE harmony_task SET owner=NULL WHERE id=$1`, w.ID)
+				if err != nil {
+					logger.Error("Cannot remove self from owner field: ", err)
+					continue // not really fatal, but not great
+				}
+			}
+			if !h.considerWork([]TaskID{TaskID(w.ID)}) {
+				logger.Error("Strange: Unable to accept previously owned task: ", w.ID, w.Name)
+			}
+		}
+	}
+	for _, h := range e.handlers {
+		go h.Adder(h.AddTask)
+	}
+	go e.poller()
+
+	return e, nil
+}
+
+// GracefullyTerminate hangs until all present tasks have completed.
+// Call this to cleanly exit the process. As some processes are long-running,
+// passing a deadline will ignore those still running (to be picked-up later).
+func (e *TaskEngine) GracefullyTerminate(deadline time.Duration) {
+	e.grace()
+	e.reg.Shutdown()
+	deadlineChan := time.NewTimer(deadline).C
+
+	// TODO block bumps & follows by unreg from DBs.
+	_, err := e.db.Exec(context.Background(), `DELETE FROM harmony_task_impl WHERE owner_id=$1`, e.ownerID)
+	if err != nil {
+		logger.Warn("Could not clean-up impl table: %w", err)
+	}
+	_, err = e.db.Exec(context.Background(), `DELETE FROM harmony_task_follow WHERE owner_id=$1`, e.ownerID)
+	if err != nil {
+		logger.Warn("Could not clean-up impl table: %w", err)
+	}
```

Comment:
Any side effects if one passes and the second fails? Would we want this to be a transaction or is it better if at least one passes?

Reply:
It best to clean up these tables as much as possible since they become serial HTTP requests later on. It's a best-effort design, so excess entries don't break things too much: we just fall back to the slow way if it takes over 1 minute.